base/util/PKICertImport.bash

```diff
@@ -258,36 +495,322 @@ function PKICertImport() {
+
+        # Arguments for listing certificates.
+        local list_args=()
+        list_args+=("-d" "$NSSDB_TYPE$NSSDB")
+
+        # Arguments for import
+        local import_args=("-i" "$CERT_PATH")
+        import_args+=("-d" "$NSSDB_TYPE$NSSDB")
+
+        # When present, specify NSS DB password for listing and importing.
+        if [ "x$NSSDB_PASSWORD" != "x" ]; then
+            list_args+=("-f" "$NSSDB_PASSWORD")
+            import_args+=("-k" "$NSSDB_PASSWORD")
+        fi
+
+        # When present, specify .p12 file password on import.
+        if [ "x$PKCS12_PASSWORD" != "x" ]; then
+            import_args+=("-w" "$PKCS12_PASSWORD")
+        fi
+
+        # List certificates and keys prior to import.
+        __v certutil -L "${list_args[@]}"
+        __v certutil -K "${list_args[@]}"
+        certutil -L "${list_args[@]}" |
+            grep "[$tf]*,[$tf]*,[$tf]*" |
+            sed "s/[[:space:]]*[$tf]*,[$tf]*,[$tf]*//g" |
+            sed 's/[[:space:]]*$//g' |
+            sort > "$before_certs"
+        certutil -K -d dbs/verify/ |
+            grep '^<[[:space:]]*[0-9]*>' |
+            sed 's/^<[[:space:]]*[0-9]*>[[:space:]]*rsa//g' |
+            awk '{print $1}' |
+            sort > "$before_keys"
+
+        # Perform import
+        __v pk12util "${import_args[@]}"
+        pk12util "${import_args[@]}"
+        ret=$?
+
+        if (( ret != 0 )); then
+            __e "Error importing certificates from .p12 file!"
+            _remove_cert "$CERT_NICKNAME"
+            __secure_rmtmp
+            exit 1
+        fi
+
+        # List certificates and keys after import.
+        __v certutil -L "${list_args[@]}"
+        __v certutil -K "${list_args[@]}"
+        certutil -L "${list_args[@]}" |
+            grep "[$tf]*,[$tf]*,[$tf]*" |
+            sed "s/[[:space:]]*[$tf]*,[$tf]*,[$tf]*//g" |
+            sed 's/[[:space:]]*$//g' |
+            sort > "$after_certs"
+        certutil -K -d dbs/verify/ |
+            grep '^<[[:space:]]*[0-9]*>' |
+            sed 's/^<[[:space:]]*[0-9]*>[[:space:]]*rsa//g' |
+            awk '{print $1}' |
+            sort > "$after_keys"
+
+        mapfile -t PKCS12_NODES < <(comm -13 "$before_certs" "$after_certs")
+        mapfile -t PKCS12_KEYS < <(comm -13 "$before_keys" "$after_keys")
+    }
+
+    # Verify leaf nodes in a chain when --chain is specified. Also handles
+    # specifiying turst on a certificate.
+    function _verify_chain() {
+        local remaining_nodes=("${PKCS12_NODES[@]}")
+
+        # Trust certificate before validation -- UNSAFE flag
+        if [ "$PKCS12_UNSAFE" == "true" ]; then
+            for cert in "${PKCS12_NODES[@]}"; do
+                _trust_cert "$cert" "$PKCS12_CHAIN_TRUST"
+            done
+        fi
+
+        # If we don't need to validate anything, go ahead and exit
+        if [ "$PKCS12_CHAIN" == "false" ]; then
+            return 0
+        fi
+
+        pre_count=${#remaining_nodes}
+        while (( pre_count > 0 )); do
```

Comment:
Since our certs aren't in order, we have to loop until we're either done or can't make any more progress. Errors aren't fatal until we fail to validate all certs in the chain.